File: Java/src/main/java/org/ulpgc/utils/MatrixGenerator.java
package org.ulpgc.utils;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.nio.file.StandardOpenOption;
import java.util.Random;

public class MatrixGenerator {

    private static final Random RANDOM = new Random();

    public static double[][] generateDenseMatrix(int rows, int cols, double minValue, double maxValue) {
        validateParameters(rows, cols, minValue, maxValue);
        double[][] matrix = new double[rows][cols];
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                matrix[i][j] = getRandomValue(minValue, maxValue);
            }
        }
        return matrix;
    }

    public static void generateMatrixForMapReduce(int rows, int cols, String matrixType, double minValue, double maxValue, String filePath) {
        validateParameters(rows, cols, minValue, maxValue);
        StringBuilder builder = new StringBuilder();

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                double value = getRandomValue(minValue, maxValue);
                builder.append(matrixType).append(",").append(i).append(",").append(j).append(",").append(value).append("\n");
            }
        }

        try {
            Files.createDirectories(Paths.get(filePath).getParent());
            Files.write(Paths.get(filePath), builder.toString().getBytes(), StandardOpenOption.CREATE, StandardOpenOption.TRUNCATE_EXISTING);
            System.out.println("Matrix " + matrixType + " saved to: " + filePath);
        } catch (IOException e) {
            System.err.println("Error writing matrix to file: " + e.getMessage());
        }
    }

    private static double getRandomValue(double minValue, double maxValue) {
        return minValue + (maxValue - minValue) * RANDOM.nextDouble();
    }

    private static void validateParameters(int rows, int cols, double minValue, double maxValue) {
        if (rows <= 0 || cols <= 0) {
            throw new IllegalArgumentException("Rows and columns must be greater than 0.");
        }
        if (minValue >= maxValue) {
            throw new IllegalArgumentException("minValue must be less than maxValue.");
        }
    }
}

File: Python/utils/matrix_generator.py
import os
import numpy as np

class MatrixGenerator:
    @staticmethod
    def generate_dense_matrix(rows, cols, min_value=1, max_value=10):
        if rows <= 0 or cols <= 0:
            raise ValueError("Rows and columns must be greater than 0.")
        return np.random.uniform(min_value, max_value, (rows, cols))

    @staticmethod
    def generate_matrix_for_mapreduce(rows, cols, matrix_type, min_value, max_value, file_path):
        os.makedirs(os.path.dirname(file_path), exist_ok=True)

        with open(file_path, "w") as file:
            for i in range(rows):
                for j in range(cols):
                    value = np.random.uniform(min_value, max_value)
                    file.write(f"{matrix_type},{i},{j},{value}\n")
        print(f"Matrix {matrix_type} saved to: {file_path}")
